cmd/crane/cmd/push.go

```diff
@@ -15,26 +15,74 @@
 package cmd
 
 import (
+	"errors"
 	"fmt"
+	"os"
 
 	"github.com/google/go-containerregistry/pkg/crane"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
 	"github.com/spf13/cobra"
 )
 
 // NewCmdPush creates a new cobra.Command for the push subcommand.
 func NewCmdPush(options *[]crane.Option) *cobra.Command {
-	return &cobra.Command{
-		Use:   "push TARBALL IMAGE",
-		Short: "Push image contents as a tarball to a remote registry",
+	cmd := &cobra.Command{
+		Use:   "push PATH IMAGE",
+		Short: "Push local image contents to a remote registry",
+		Long:  `If the PATH is a directory, it will be read as an OCI image layout. Otherwise, PATH is assumed to be a docker-style tarball.`,
 		Args:  cobra.ExactArgs(2),
 		RunE: func(_ *cobra.Command, args []string) error {
 			path, tag := args[0], args[1]
-			img, err := crane.Load(path)
+
+			img, err := loadImage(path)
 			if err != nil {
-				return fmt.Errorf("loading %s as tarball: %w", path, err)
+				return err
 			}
 
 			return crane.Push(img, tag, *options...)
 		},
 	}
+	return cmd
+}
+
+func loadImage(path string) (v1.Image, error) {
+	dir, err := isDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	if !dir {
+		img, err := crane.Load(path)
+		if err != nil {
+			return nil, fmt.Errorf("loading %s as tarball: %w", path, err)
+		}
+		return img, nil
+	}
+
+	l, err := layout.ImageIndexFromPath(path)
+	if err != nil {
+		return nil, fmt.Errorf("loading %s as OCI layout: %w", path, err)
+	}
+	m, err := l.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+	if len(m.Manifests) != 1 {
+		return nil, errors.New("pushing multiple images from layout not yet supported (PRs welcome)")
+	}
+	return l.Image(m.Manifests[0].Digest)
```

Comment:
Haven't you marked this as `Fixes: #{the issue covering pushing layouts}`?

Maybe we should leave one open

Reply:
There are ~3 ways I can imagine people wanting to use this, let's chat.

Reply:
Did we resolve the next steps here?  I'm fine with landing this part as-is, assuming it's good enough for `kaniko`'s `--oci-layout`.  However, I'd like a clear path for how we support the same for `ko`.